planner/core/rule_aggregation_elimination.go

```diff
@@ -155,3 +272,174 @@ func (a *aggregationEliminator) optimize(p LogicalPlan) (LogicalPlan, error) {
 	}
 	return p, nil
 }
+
+func collectNestedAggPattern(lp LogicalPlan, ptn *nestedAggPattern) {
+	if len(lp.Children()) != 1 {
+		return
+	}
+	child := lp.Children()[0]
+	switch v := child.(type) {
+	case *LogicalAggregation:
+		ptn.inner = v
+		return
+	case *LogicalProjection:
+		if ptn.proj != nil {
+			return
+		}
+		ptn.proj = v
+	case *LogicalSort:
+		// Ignore sort between aggregations.
+	default:
+		ptn.isTrivial = false
+	}
+	collectNestedAggPattern(child, ptn)
+}
+
+// tryToCombineAggFunc checks the types of inner/outer aggregate function and check if they can be combined as one based on their semantics.
+// for example, since max(max(PARTIAL)) can be combined as max(TOTAL), we can combine inner max() and outer max() as a final max()
+// `innerIsGroup` indicates if the inner aggregate function is aggregating group-by items(values are de-duplicated).
+func tryToCombineAggFunc(outer, inner *aggregation.AggFuncDesc, innerIsGroup bool, expr expression.Expression) *aggregation.AggFuncDesc {
+	combined := outer.Clone()
+	combined.Args[0] = expr
+
+	if innerIsGroup {
+		if (inner.Name == ast.AggFuncFirstRow || inner.Name == ast.AggFuncMax || inner.Name == ast.AggFuncMin) &&
```

Comment:
Could you explain the rationale behind this condition? From my understanding, if `innerIsGroup` is true, we can just keep the outer aggregation without change, no matter what kind of aggregation it is?

Reply:
If the item being aggregated in outter aggregation is group-by item of inner aggregation. We can only do elimination on very few conditions.
For example, given
```
select sum(at) from (select count(a) as at, b as bt from t group by a, b) as t group by bt
```
AggFunc `count(a)` is always 1 and `sum(at)` is distinct number of `a`. We only consider the simplest case here because for other cases, there is no common rule to indicate the semantics.